scipy/linalg/basic.py

```diff
@@ -1293,7 +1297,9 @@ def pinv(a, cond=None, rcond=None, return_rank=False, check_finite=True):
     """
     a = _asarray_validated(a, check_finite=check_finite)
     b = np.identity(a.shape[0], dtype=a.dtype)
-    if rcond is not None:
+    if rcond is None:
+        cond = max(a.shape) * np.spacing(a.real.dtype.type(1))
+    else:
         cond = rcond
```

Comment:
If I do:
```
pinv(a, 0.1)
```
my `cond=0.1` value will end up not being used because this code path will enter the `if rcond is None` part of the branch, and overwrite `cond`, right?

If so, we need some more tests that would have caught this sort of thing. I'd do a `pytest.mark.parametrize` over `cond, rcond` pairs operating on an array that, if the specified condition is not respected, will not `np.allclose` with an expected result. To make a very simple example with `0.1` as the potential `cond/rcond` and the expected result `[[1, 0], [0, 0]]` when it is used properly (and `np.eye(2)` when it is not):
```
>>> from scipy.linalg import pinv
>>> import numpy as np
>>> x = np.array([[1, 0], [0, 0.01]])
>>> pinv(x)
array([[  1.,   0.],
       [  0., 100.]])
>>> pinv(x, 0.1)
array([[1., 0.],
       [0., 0.]])
```

Reply:
Yes indeed here I didn't touch the original flow but merely inverted it. To be honest, I am not entirely sure what these keywords meant to do. And also why there are two of them (cond and rcond) since one is supposed to be the `r`eciprocal of the other but in the code there is no such distinction. Hence I didn't want to change the behavior but we can actually change it in a separate PR now that you also noticed it. 

But your test idea is a good one. I will try to find something that fails on master.

Reply:
> Yes indeed here I didn't touch the original flow but merely inverted it.

Assuming by this you mean that the rcond/cond conditional behavior is unchanged (i.e., only what actually gets computed is changed, when it's necessary to do so), I think that's not the case currently. On `master` the conditional implies a truth table like:

- `rcond is None`
  - `cond is None`: compute (presumably? Both remain `None` so I assume it's computed somewhere later)
  - `cond is not None`: use `cond`
- `rcond is not None`
  - `cond is None`: use `rcond`
  - `cond is not None`: use `rcond`

In other words, it was a "prefer rcond, and if neither is provided, compute" mode.

On this PR, this becomes:

- `rcond is None`
  - `cond is None`: compute
  - `cond is not None`: **compute**
- `rcond is not None`
  - `cond is None`: use `rcond`
  - `cond is not None`: use `rcond`

This becomes a "rcond is used if it's not None, else compute (cond is ignored)". In other words, the `rcond is None, cond is not None` path has been broken. Or do I miss something here?

Reply:
(FWIW I usually see `rcond` as a *relative* (to max singular value) condition, and `cond` as an *absolute* condition, but that doesn't seem to be how it was used here.)

Reply:
I don't know where it is actually converted. This is the current master code 

https://github.com/scipy/scipy/blob/master/scipy/linalg/basic.py#L1296

and there is no conversion done anywhere else since it is passed directly to `scipy.linalg.lstsq`. I am even more confused now 😃 I think we botched this up at some point.

Reply:
Ah now I see what you mean.